### ISSUE TITLE
Captured object: why stderr will shown on first usage instead of after execution?

### ISSUE THREAD

**Issue by anki-code** [xonsh/xonsh#4220]
Hi!

I expect that captured object will show the errors after execution but it show stderr only after first using.
```python
r = !(echo err o>e)
# Expected: err
# Result: empty

bool(r)
# err
# True
```

I've found a test around this logic:

<details>

https://github.com/xonsh/xonsh/blob/5a792a6eec4889cbbc50619da6b26f7471d7ce5d/tests/test_integrations.py#L163-L176

</details>

And the actual code that shows stderr:

<details>

https://github.com/xonsh/xonsh/blob/5a792a6eec4889cbbc50619da6b26f7471d7ce5d/xonsh/procs/pipelines.py#L399-L404

</details>

And the `run_subproc` command does `end` but not for object:

<details>

https://github.com/xonsh/xonsh/blob/5a792a6eec4889cbbc50619da6b26f7471d7ce5d/xonsh/procs/specs.py#L876-L887

</details>

As result I've found that there's no execution of the `.end()` event after execution.

This is bug or feature or workaround for something?

Is this expected behavior: to show stderr on the first usage of the object instead of after executing?

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by daniel-shimon** [comment]
The thing is the command's stdout and stderr are read lazily when either of them is needed.
That's a feature and expected. The real issue here is that **stderr isn't being captured by `!()`**, and is printed out lazily instead

**Comment by daniel-shimon** [comment]
Closed by #4221 and #4231
